expression-language.md

```diff
@@ -0,0 +1,369 @@
+- 1.2. [Relation to Subscriptions API](#12-relation-to-the-subscriptions-api)
+
+2. [Language syntax](#2-language-syntax)
+
+- 2.1. [Expression](#21-expression)
+- 2.2. [Value identifiers and literals](#22-value-identifiers-and-literals)
+- 2.3. [Operators](#23-operators)
+- 2.4. [Functions invocation](#24-functions-invocation)
+
+3. [Language semantics](#3-language-semantics)
+
+- 3.1. [Type System](#31-type-system)
+- 3.2. [CloudEvent context identifiers and types](#32-cloudevent-context-identifiers-and-types)
+- 3.3. [Errors](#33-errors)
+- 3.4. [Operators](#34-operators)
+- 3.5. [Functions](#35-functions)
+- 3.6. [Evaluation of the expression](#36-evaluation-of-the-expression)
+
+4. [Examples](#4-examples)
+5. [References](#5-references)
+
+## 1. Introduction
+
+CloudEvents SQL expressions (also known as CESQL) allow matching of CloudEvent attributes against complex expressions
+that lean on the syntax of Structured Query Language (SQL) `WHERE` clauses. Using SQL-derived expressions for message
+filtering is in widespread implementation use because the Java Message Service (JMS) message selector syntax also leans
+on SQL. Note that neither the SQL standard (ISO 9075) nor the JMS standard nor any other SQL dialect are used as a
+normative foundation or constrain the expression syntax defined in this specification, but the syntax is informed by
+them.
+
+CESQL is a _[Total pure functional programming language][total-programming-language-wiki]_ in order to guarantee the
+termination of the evaluation of the expression. It features a type system correlated to the [CloudEvents type
+system][ce-type-system] and it features boolean and arithmetics operations, as well as built-in functions for string
+manipulation.
+
+The language is not constrained to a particular execution environment, which means it might run in a source, in a
+producer, in an intermediary, and it can be implemented using any tech stack.
+
+The CloudEvents Expression Language assumes the input always includes, but it's not limited to, a single valid and type
+checked CloudEvent instance. An expression cannot mutate the value of the input CloudEvent instance, nor any of the
+other input values. The evaluation of an expression observes the concept of [referential
+transparency][referential-transparency-wiki].
+
+The CloudEvents Expression Language doesn't support the handling of the data field of the CloudEvent instances, due to
+its polymorphic nature and complexity. We strongly encourage users that needs this functionality to use other more
+appropriate tools.
+
+### 1.1. Conformance
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and
+"OPTIONAL" in this document are to be interpreted as described in [RFC2119][rfc2119].
+
+### 1.2. Relation to the Subscriptions API
+
+The CESQL can be used as a [filter dialect][subscriptions-filter-dialect] to filter out the input values
+
+<!-- TODO -->
```

Comment:
I would love to revisit this paragraph after we merge this PR, in order to align with Subscription api (which I guess will need to contain a little text about this filter dialect)